### PR TITLE
S3 Gateway: Check for encryption headers properly

### PR DESCRIPTION
**Pull request by kannappanr**
## Description
encryption functional tests in minio-go are failing against S3 Gateway as key rotation checks are not done correctly. This PR checks if the source is encrypted or the target will be encrypted.

## Motivation and Context
minio-go encryption tests are failing against s3 gateway

## How to test this PR?
Run minio-go functional tests against S3 gateway

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [ ] Unit tests needed

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :x: [more...](#8708aa96-7842-4b42-a4e0-396caa59bcc0) |
| mint-gateway-azure.sh | :x: [more...](#c35a6ed2-f6d4-4592-82b0-b3e4d2ab9950) |
| mint-erasure.sh | :x: [more...](#c701386f-5c6b-4ffc-9cbc-a3fd3973386d) |
| mint-zoned.sh | :x: [more...](#6c28e959-b34a-43d0-af8f-a69f748289f0) |
<a name="6c28e959-b34a-43d0-af8f-a69f748289f0"></a>
### 10309-5646a6a/mint-zoned.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:31155
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp c78098a01426:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 2 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 43 seconds
(4/15) Running aws-sdk-ruby tests ... done in 5 seconds
(5/15) Running awscli tests ... done in 1 minutes and 19 seconds
(6/15) Running healthcheck tests ... done in 1 seconds
(7/15) Running mc tests ... done in 29 seconds
(8/15) Running minio-dotnet tests ... done in 43 seconds
(9/15) Running minio-go tests ... done in 2 minutes and 22 seconds
(10/15) Running minio-java tests ... done in 1 minutes and 28 seconds
(11/15) Running minio-js tests ... done in 55 seconds
(12/15) Running minio-py tests ... done in 3 minutes and 16 seconds
(13/15) Running s3cmd tests ... FAILED in 18 seconds
{
  "name": "s3cmd",
  "duration": "8807",
  "function": "test_sync_list_objects",
  "status": "FAIL",
  "error": "WARNING: Bucket is not empty. Removing all the objects from it first. This may take some time...\nERROR: S3 error: 404 (Not Found)"
}
(13/15) Running s3select tests ... done in 7 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="8708aa96-7842-4b42-a4e0-396caa59bcc0"></a>
### 10309-5646a6a/mint-gateway-s3.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev7.minio.io:31368
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 3a317297de6a:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 41 seconds
(4/15) Running aws-sdk-ruby tests ... done in 2 seconds
(5/15) Running awscli tests ... done in 2 minutes and 1 seconds
(6/15) Running healthcheck tests ... done in 1 seconds
(7/15) Running mc tests ... done in 49 seconds
(8/15) Running minio-dotnet tests ... done in 36 seconds
(9/15) Running minio-go tests ... FAILED in 10 minutes and 53 seconds
{
  "args": {
    "destination": {
      "Bucket": "minio-go-test-uc30fbcfh46q3tuq",
      "Object": "srcObject",
      "Encryption": {},
      "UserMetadata": null,
      "ReplaceMetadata": false,
      "UserTags": null,
      "ReplaceTags": false,
      "LegalHold": "",
      "Mode": "",
      "RetainUntilDate": "0001-01-01T00:00:00Z",
      "Size": 0,
      "Progress": null
    },
    "source": {
      "Bucket": "minio-go-test-uc30fbcfh46q3tuq",
      "Object": "srcObject",
      "VersionID": "",
      "MatchETag": "",
      "NoMatchETag": "",
      "MatchModifiedSince": "0001-01-01T00:00:00Z",
      "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
      "MatchRange": false,
      "Start": 0,
      "End": 0,
      "Encryption": null
    }
  },
  "duration": 600471,
  "error": "Put \"http://minio-dev7.minio.io:31368/minio-go-test-uc30fbcfh46q3tuq/srcObject\": net/http: timeout awaiting response headers",
  "function": "CopyObject(destination, source)",
  "message": "CopyObject failed",
  "name": "minio-go: testEncryptedSSES3ToSSES3CopyObject",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 25 seconds
(10/15) Running minio-js tests ... done in 39 seconds
(11/15) Running minio-py tests ... done in 1 minutes and 7 seconds
(12/15) Running s3cmd tests ... done in 58 seconds
(13/15) Running s3select tests ... done in 2 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="c35a6ed2-f6d4-4592-82b0-b3e4d2ab9950"></a>
### 10309-5646a6a/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev7.minio.io:31899
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 946cf43cdce5:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 7 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 4 minutes and 41 seconds
(4/15) Running aws-sdk-ruby tests ... done in 17 seconds
(5/15) Running awscli tests ... done in 2 minutes and 39 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 2 minutes and 50 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 33 seconds
(9/15) Running minio-go tests ... FAILED in 5 minutes and 4 seconds
{
  "args": {},
  "duration": 932,
  "error": "The specified multipart upload does not exist. The upload ID may be invalid, or the upload may have been aborted or completed.",
  "function": "CopyObjectPart(destination, source)",
  "message": "CompleteMultipartUpload call failed",
  "name": "minio-go: testUnencryptedToSSES3CopyObjectPart",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 5 minutes and 44 seconds
(10/15) Running minio-js tests ... done in 2 minutes and 35 seconds
(11/15) Running minio-py tests ... done in 9 minutes and 53 seconds
(12/15) Running s3cmd tests ... done in 1 minutes and 20 seconds
(13/15) Running s3select tests ... done in 1 minutes and 1 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
<a name="c701386f-5c6b-4ffc-9cbc-a3fd3973386d"></a>
### 10309-5646a6a/mint-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev8.minio.io:31775
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 57f258ef1123:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 42 seconds
(4/15) Running aws-sdk-ruby tests ... done in 2 seconds
(5/15) Running awscli tests ... done in 2 minutes and 9 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 50 seconds
(8/15) Running minio-dotnet tests ... done in 39 seconds
(9/15) Running minio-go tests ... FAILED in 16 minutes and 45 seconds
{
  "args": {
    "bucketName": "minio-go-test-ty9mpp6lhnhk3pa5",
    "fileName": "/mint/data/datafile-129-MB",
    "location": "us-east-1",
    "objectName": "minio-go-test-ty9mpp6lhnhk3pa5FPutObject-GTar",
    "opts": {
      "UserMetadata": null,
      "UserTags": null,
      "Progress": null,
      "ContentType": "",
      "ContentEncoding": "",
      "ContentDisposition": "",
      "ContentLanguage": "",
      "CacheControl": "",
      "Mode": "",
      "RetainUntilDate": "0001-01-01T00:00:00Z",
      "ServerSideEncryption": null,
      "NumThreads": 0,
      "StorageClass": "",
      "WebsiteRedirectLocation": "",
      "PartSize": 0,
      "LegalHold": "",
      "SendContentMd5": false,
      "DisableMultipart": false,
      "ReplicationVersionID": "",
      "ReplicationStatus": "",
      "ReplicationMTime": "0001-01-01T00:00:00Z"
    }
  },
  "duration": 967389,
  "error": "The specified bucket does not exist",
  "function": "FPutObject(bucketName, objectName, fileName, opts)",
  "message": "FPutObject failed",
  "name": "minio-go: testFPutObject",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 42 seconds
(10/15) Running minio-js tests ... done in 48 seconds
(11/15) Running minio-py tests ... done in 2 minutes and 11 seconds
(12/15) Running s3cmd tests ... done in 21 seconds
(13/15) Running s3select tests ... done in 3 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

File: cmd/crypto/header.go [hunk 1]
```diff
@@ -94,6 +94,12 @@ func IsRequested(h http.Header) bool {
 	return S3.IsRequested(h) || SSEC.IsRequested(h) || SSECopy.IsRequested(h) || S3KMS.IsRequested(h)
 }
 
+// IsEncryptionRequested checks if encryption is
+// requested.
+func IsEncryptionRequested(h http.Header) bool {
+	return S3.IsRequested(h) || SSEC.IsRequested(h)
+}
```
Comment (poornas) [inline cmd/crypto/header.go]: is this function required? IsRequested should suffice

**Review by poornas — CHANGES_REQUESTED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :x: [more...](#619c1dc2-49f6-41c5-8af4-d1738083605a) |
| mint-gateway-azure.sh | :x: [more...](#6c1509e8-ef36-4f16-9652-aaa1612979e3) |
<a name="6c1509e8-ef36-4f16-9652-aaa1612979e3"></a>
### 10309-949355f/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev4.minio.io:31792
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 4f74b1f2dc59:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 7 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 4 minutes and 46 seconds
(4/15) Running aws-sdk-ruby tests ... done in 18 seconds
(5/15) Running awscli tests ... done in 3 minutes and 39 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 3 minutes and 12 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 32 seconds
(9/15) Running minio-go tests ... FAILED in 1 minutes and 45 seconds
{
  "args": {
    "bucketName": "minio-go-test-dfr0xd2sbv03mkfq",
    "objectName": "test-object",
    "opts": "",
    "size": -1
  },
  "duration": 642,
  "function": "PutObject(bucketName, objectName, reader,size,opts)",
  "message": "Unexpected size",
  "name": "minio-go: testPutObjectStreaming",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 6 minutes and 19 seconds
(10/15) Running minio-js tests ... FAILED in 46 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 75,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:86:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:678:10) at endReadableNT (_stream_readable.js:1244:12) at processTicksAndRejections (internal/process/task_queues.js:80:21)"
}
(10/15) Running minio-py tests ... done in 12 minutes and 46 seconds
(11/15) Running s3cmd tests ... done in 1 minutes and 25 seconds
(12/15) Running s3select tests ... done in 33 seconds
(13/15) Running security tests ... done in 0 seconds

Executed 13 out of 15 tests successfully.
```
<a name="619c1dc2-49f6-41c5-8af4-d1738083605a"></a>
### 10309-949355f/mint-gateway-s3.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev7.minio.io:31403
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 220839b33de2:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 41 seconds
(4/15) Running aws-sdk-ruby tests ... done in 2 seconds
(5/15) Running awscli tests ... done in 2 minutes and 0 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 36 seconds
(8/15) Running minio-dotnet tests ... done in 30 seconds
(9/15) Running minio-go tests ... FAILED in 11 minutes and 3 seconds
{
  "args": {
    "destination": {
      "Bucket": "minio-go-test-jd691p9gfzfzyqaj",
      "Object": "srcObject",
      "Encryption": {},
      "UserMetadata": null,
      "ReplaceMetadata": false,
      "UserTags": null,
      "ReplaceTags": false,
      "LegalHold": "",
      "Mode": "",
      "RetainUntilDate": "0001-01-01T00:00:00Z",
      "Size": 0,
      "Progress": null
    },
    "source": {
      "Bucket": "minio-go-test-jd691p9gfzfzyqaj",
      "Object": "srcObject",
      "VersionID": "",
      "MatchETag": "",
      "NoMatchETag": "",
      "MatchModifiedSince": "0001-01-01T00:00:00Z",
      "MatchUnmodifiedSince": "0001-01-01T00:00:00Z",
      "MatchRange": false,
      "Start": 0,
      "End": 0,
      "Encryption": null
    }
  },
  "duration": 601157,
  "error": "Put \"http://minio-dev7.minio.io:31403/minio-go-test-jd691p9gfzfzyqaj/srcObject\": net/http: timeout awaiting response headers",
  "function": "CopyObject(destination, source)",
  "message": "CopyObject failed",
  "name": "minio-go: testEncryptedSSES3ToSSES3CopyObject",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 19 seconds
(10/15) Running minio-js tests ... done in 33 seconds
(11/15) Running minio-py tests ... done in 1 minutes and 1 seconds
(12/15) Running s3cmd tests ... done in 59 seconds
(13/15) Running s3select tests ... done in 2 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally
Error: No such image: minio/minio:10309-949355f

File: cmd/crypto/header.go [hunk 2]
```diff
@@ -94,6 +94,12 @@ func IsRequested(h http.Header) bool {
 	return S3.IsRequested(h) || SSEC.IsRequested(h) || SSECopy.IsRequested(h) || S3KMS.IsRequested(h)
 }
 
+// IsEncryptionRequested checks if encryption is
+// requested.
+func IsEncryptionRequested(h http.Header) bool {
```
Comment (harshavardhana) [inline cmd/crypto/header.go]: This is not needed @kannappanr crypto.IsRequested() is enough

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/crypto/header.go (hunk 2 above)
Comment (kannappanr) [reply cmd/crypto/header.go]: done

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#e37a1be5-8bdf-4be4-8451-2607489882a1) |
<a name="e37a1be5-8bdf-4be4-8451-2607489882a1"></a>
### 10309-6df58bb/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c3.minio.io:30035
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 96c2b9a3da5e:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 7 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 4 minutes and 45 seconds
(4/15) Running aws-sdk-ruby tests ... done in 16 seconds
(5/15) Running awscli tests ... done in 2 minutes and 38 seconds
(6/15) Running healthcheck tests ... done in 0 seconds
(7/15) Running mc tests ... done in 4 minutes and 20 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 31 seconds
(9/15) Running minio-go tests ... FAILED in 5 minutes and 15 seconds
{
  "args": {},
  "duration": 1272,
  "error": "The specified multipart upload does not exist. The upload ID may be invalid, or the upload may have been aborted or completed.",
  "function": "CopyObjectPart(destination, source)",
  "message": "CompleteMultipartUpload call failed",
  "name": "minio-go: testUnencryptedToSSES3CopyObjectPart",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 5 minutes and 55 seconds
(10/15) Running minio-js tests ... done in 2 minutes and 35 seconds
(11/15) Running minio-py tests ... done in 10 minutes and 22 seconds
(12/15) Running s3cmd tests ... done in 1 minutes and 31 seconds
(13/15) Running s3select tests ... done in 49 seconds
(14/15) Running security tests ... done in 0 seconds

Executed 14 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally
Error: No such image: minio/minio:10309-6df58bb

**Review by harshavardhana — APPROVED** [review]

**Pull request merged**